pkg/puller/downloader.go

```diff
@@ -0,0 +1,172 @@
+package puller
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"hash/fnv"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+var (
+	d *Downloader
+)
+
+type Downloader struct {
+	NumRetries int
+	ModelDir   string
+}
+
+type Protocol string
+
+const (
+	S3    Protocol = "s3://"
+	GCS    Protocol = "gs://"
+	//PVC   Protocol = "pvc://"
+	//File  Protocol = "file://"
+	//HTTPS Protocol = "https://"
+)
+
+// TODO: Add more supported protocols
+var SupportedProtocols = []Protocol{S3, GCS}
+
+func init() {
+	d = NewDownloader()
+}
+
+func NewDownloader() *Downloader {
+	d := new(Downloader)
+	return d
+}
+
+func SetDownloadSettings(numRetries int, modelDir string) {
```

Comment:
I'd probably avoid setter/ctor patterns in golang unless you need a builder/factory pattern. This is one of the nice things about go object initialization.

Just set the variables directly at creation

downloader := Downloader {
 NumRetries: ...
 ModelDir: ...
}